reframe/core/schedulers/sge.py

```diff
@@ -0,0 +1,162 @@
+        jobid_match = re.search(r'^Your job (?P<jobid>\S+)', completed.stdout)
+        if not jobid_match:
+            raise JobSchedulerError('could not retrieve the job id '
+                                    'of the submitted job')
+
+        job._jobid = jobid_match.group('jobid')
+        job._submit_time = time.time()
+
+    def poll(self, *jobs):
+        if jobs:
+            # Filter out non-jobs
+            jobs = [job for job in jobs if job is not None]
+
+        if not jobs:
+            return
+
+        user = osext.osuser()
+        completed = osext.run_command(
+            f'qstat -xml -u {user}'
+        )
+
+        if completed.returncode != 0:
+            raise JobSchedulerError(
+                f'qstat failed with exit code {completed.returncode} '
+                f'(standard error follows):\n{completed.stderr}'
+            )
+
+        root = ET.fromstring(completed.stdout)
+
+        # Store information for each job separately
+        jobinfo = {}
+        for queue_info in root:
+
+            # Reads the XML and prints jobs with status belonging to user.
+            if queue_info is None:
+                raise JobSchedulerError('Decomposition error!\n')
+
+            for job_list in queue_info:
+                if job_list.find("JB_owner").text != user:
+                    # Not a job of this user.
+                    continue
+
+                job_number = job_list.find("JB_job_number").text
+
+                if job_number not in [job.jobid for job in jobs]:
+                    # Not a reframe job.
+                    continue
+
+                state = job_list.find("state").text
+
+                # For the list of known statuses see `man 5 sge_status`
+                # (https://arc.liv.ac.uk/SGE/htmlman/htmlman5/sge_status.html)
+                if state in ['r', 'hr', 't', 'Rr', 'Rt']:
+                    jobinfo[job_number] = 'RUNNING'
+                elif state in ['qw', 'Rq', 'hqw', 'hRwq']:
+                    jobinfo[job_number] = 'PENDING'
+                elif state in ['s', 'ts', 'S', 'tS', 'T', 'tT', 'Rs',
+                               'Rts', 'RS', 'RtS', 'RT', 'RtT']:
+                    jobinfo[job_number] = 'SUSPENDED'
+                elif state in ['Eqw', 'Ehqw', 'EhRqw']:
+                    jobinfo[job_number] = 'ERROR'
+                elif state in ['dr', 'dt', 'dRr', 'dRt', 'ds',
+                               'dS', 'dT', 'dRs', 'dRS', 'dRT']:
+                    jobinfo[job_number] = 'DELETING'
+                elif state == 'z':
+                    jobinfo[job_number] = 'COMPLETED'
+
+        for job in jobs:
+            if job.jobid not in jobinfo:
+                self.log(f'Job {job.jobid} not known to scheduler, '
+                         f'assuming job completed')
+                job._state = 'COMPLETED'
+                job._completed = True
+                continue
+
+            job._state = jobinfo[job.jobid]
```

Comment:
I don't think that you need this loop here. If you keep track of the jobs submitted by reframe, you can just directly update the job state as you iterate over the XML output.